triton-cli/src/main.rs

```diff
@@ -89,171 +89,122 @@ fn parse_inputs(inputs: Option<String>) -> Vec<BFieldElement> {
+        .with_context(|| "Triton VM errored during program execution")?;
     println!("success!");
 
-    // write the proof in an arbitrary binary format
-    //
-    // version: u8
-    //
-    // security_level: u64
-    // fri_expansion_factor: u64
-    // num_trace_randomizers: u64
-    // num_colinearity_checks: u64
-    // num_combination_codeword_checks: u64
-    //
-    // digest_length: u8
-    // public_input_length: u64
-    // public_output_length: u64
-    // proof_length: u64
-    //
-    // program_digest: u64[digest_length]
-    //
-    // public_inputs: u64[public_input_length]
-    // public_outputs: u64[public_output_length]
-    // proof: u64[proof_length]
-    write_proof(data, out).expect("failed to write proof to file");
+    write_proof(data, out).with_context(|| "Failed to write proof to file")?;
     println!("proof written to: {out}");
+    Ok(())
+}
+
+#[derive(Serialize, Deserialize)]
+struct SerializedProof {
+    version: u8,
+    security_level: u64,
+    fri_expansion_factor: u64,
+    num_trace_randomizers: u64,
+    num_colinearity_checks: u64,
+    num_combination_codeword_checks: u64,
+    program_digest: Vec<u64>,
+    public_inputs: Vec<u64>,
+    public_outputs: Vec<u64>,
+    proof: Vec<u64>,
 }
 
-fn read_proof(proof_path: &String) -> Result<(Stark, Claim, Proof)> {
-    let file = fs::File::open(proof_path)?;
-    let mut reader = BufReader::new(file);
-    let version = read_u8(&mut reader);
-    assert!(version == 1, "wrong proof file version!");
+fn read_proof(proof_path: &str) -> Result<(Stark, Claim, Proof)> {
+    let mut file = fs::File::open(proof_path)?;
+    let mut proof_bytes: Vec<u8> = Vec::new();
+    file.read_to_end(&mut proof_bytes)?;
+    let serialized_proof: SerializedProof = bincode::deserialize(&proof_bytes)?;
     let stark = Stark {
-        security_level: read_usize(&mut reader),
-        fri_expansion_factor: read_usize(&mut reader),
-        num_trace_randomizers: read_usize(&mut reader),
-        num_collinearity_checks: read_usize(&mut reader),
-        num_combination_codeword_checks: read_usize(&mut reader),
+        security_level: usize::try_from(serialized_proof.security_level)?,
+        fri_expansion_factor: usize::try_from(serialized_proof.fri_expansion_factor)?,
+        num_trace_randomizers: usize::try_from(serialized_proof.num_trace_randomizers)?,
+        num_collinearity_checks: usize::try_from(serialized_proof.num_colinearity_checks)?,
+        num_combination_codeword_checks: usize::try_from(
+            serialized_proof.num_combination_codeword_checks,
+        )?,
     };
-    let digest_len = read_u8(&mut reader);
-    let input_len = read_u64(&mut reader);
-    let output_len = read_u64(&mut reader);
-    let proof_len = read_u64(&mut reader);
-    let mut d = Digest::default();
+    let mut digest = Digest::default();
     assert_eq!(
-        d.0.len(),
-        usize::from(digest_len),
+        digest.0.len(),
```

Comment:
Note that there is a `const DIGEST_LENGTH` (soon an associated `const Digest::LEN`), which seems more suitable.

Reply:
Even easier is probably to use one `Digest`s `try_from` methods directly, which will return an error if the length does not match.

```rust
let digest = Digest::try_from(&program_digest)?;
```

Reply:
This `try_from` variant isn't working for me. The `program_digest` is a `Vec<u64>`. Even if i change it to a `Vec<u8>` it isn't compatible because `Digest` wants a `[u8; 40]`. I ended up mapping the `Vec<u64>` into `Vec<BFieldElement>` and pulling a slice from that, and it's slightly more semantic.

Reply:
Ah, good point. I suppose we could add a `try_from` from `Vec<u64>` but I think the path you describe works well. :ok_hand: